### DISCUSSION TITLE
How do I make a `styled` Button that changes text color from the variants?

### DISCUSSION THREAD

**Discussion by la55u** [tamagui/tamagui#539] [Q&A] [answered]
I'm trying to make two simple variants of a button: 
- primary: purple background, white text
- secondary: white background, black text

How do I create this config via variants? Do I have to define both variants in both `styled(ButtonFrame)` and `styled(ButtonText)`? That doesn't seem to work for me. Can I change a specific property (color) in the `ButtonText` component based on a variant in the Frame component? In other words do variants propagate somehow to children components? I wouldn't mind if I have to use Ideally I'd like to use my button like this:

```jsx
<ButtonFrame variant="primary">
    <ButtonText>Hello</ButtonText>
</ButtonFrame>
```

Thank you

**Comment by NandoMB** [comment, marked as answer]
Ok... You problem isn't so difficult to solve. Firstly we need to understand some aspects of Tamagui. They only have 2 "headless" components: Text and Stack. You can use it from **@tamagui/core**. The rest, all of them, came from 'tamagui' package and they are not "headless". Why I'm I saying this? Because I struggle too much to understand in deep how the override works. It's a little bit pain but... We can deal with it! So let's get started!

This is the result:
![IMG_216B03169EF3-1](https://user-images.githubusercontent.com/5796053/225780327-99ddf1bc-580d-4d33-b484-0ede4e854cf9.jpeg)

First of all, we'll create a styled button with a default style and two variants as you asked for.
```tsx
// Button.styled.tsx
import { Button, styled, GetProps } from 'tamagui';

export const StyledButton = styled(Button, {
  name: 'StyledButton',
  fontSize: 16,
  fontWeight: '400',
  color: '#FFF',
  backgroundColor: '#7F7F7F',
  variants: {
    variant: {
      primary: {
        color: 'white',
        backgroundColor: 'purple'
      },
      secondary: {
        color: 'black',
        backgroundColor: 'white'
      }
    }
  }
});

export type StyledButtonProps = GetProps<typeof StyledButton>
```

Now, I'll get the custom styled button and wrap it into a component responsible for the "business" rules...


```tsx
// Button.tsx
import { StyledButton, StyledButtonProps } from './Button.styled';

type ButtonProps = StyledButtonProps & {
  // You can add more properties here...
};

export default function Button({ children, ...rest }: ButtonProps) {
  return (
    <StyledButton {...rest}>
      {children}
    </StyledButton>
  );
}
```

After that, you are able to use your new component into your screens, like this:

```tsx
import Button from '../../components/Button'; // Our custom Button
import View from '../../components/View'; // Custom Stack component

export default function Docs() {
  return (
    <View flex={1} justifyContent="center" alignItems="center" gap={16}>
      <Button>Default</Button>
      <Button variant="primary">Primary</Button>
      <Button variant="secondary">Secondary</Button>
    </View>
  )
}

```

In case a lot of warnings saying:
`
WARN  ⚠️ You passed the value "$backgroundHover" to the style property "backgroundColor", but there's no theme or token with the key "$backgroundHover". Using theme "undefined".
`
It will be necessary to implement mandatory tokens into your custom theme. You can import from **@tamagui/themes** or create your own. You can uncomment on demmand...

```tsx
import { createTheme } from 'tamagui';
import { plainTokens as tokens } from '../tokens'; // your custom tokens, or you can import from @tamagui/tokens

export const plainTheme = {
  // Default light theme example
  //
  // background: '#fff',
  // backgroundHover: tokens.color.gray2,
  // backgroundPress: tokens.color.gray4,
  // backgroundFocus: tokens.color.gray5,
  // backgroundTransparent: tokens.color.grayA1,
  // borderColor: tokens.color.gray4,
  // borderColorHover: tokens.color.gray6,
  // borderColorPress: tokens.color.gray12,
  // borderColorFocus: tokens.color.gray11,
  // color: tokens.color.gray10,
  // colorHover: tokens.color.gray9,
  // colorPress: tokens.color.gray8,
  // colorFocus: tokens.color.gray8,
  // shadowColor: tokens.color.grayA4,
  // shadowColorHover: tokens.color.grayA6,
  // shadowColorPress: tokens.color.grayA8,
  // shadowColorFocus: tokens.color.grayA8,
};

export default createTheme(plainTheme);
```

I wish I could help!

**Reply by la55u** [reply]
Thank you for the detailed answer! At the time I opened the discussion Button could not be used with styled, hencr the confusion.

**Reply by NandoMB** [reply]
Ohh, make sense... Even that, I'm happy to help you. Let me know if you have questions about the code. 🦾

**Reply by raymondctc** [reply]
Hi, I tried your method, but I don't know how to fix the Typescript error in IDE. When I try to create a component like this:

`<Button variant="primary">Primary</Button>`, IDE will highlight the `variant` props and show:

`TS2322: Type 'primary' is not assignable to type 'outlined` 

🤔
